### PR TITLE
[mlir] Fix bugs in expand_shape patterns after semantics changes

### PR DESCRIPTION
**Pull request by Max191**
After the `output_shape` field was added to `expand_shape` ops, dynamically sized expand shapes are now possible, but this was not accounted for in the folder. This PR tightens the constraints of the folder to fix this.

### CODE REVIEW

File: mlir/test/Dialect/Tensor/canonicalize.mlir [hunk 1]
```diff
@@ -1152,7 +1152,60 @@ func.func @fold_collapse_of_expand_dynamic(%arg0 : tensor<?x?xf32>, %arg1: index
...
+  %0 = tensor.collapse_shape %arg0 [[0, 1], [2]]
+      : tensor<?x?x?xf32> into tensor<?x?xf32>
+  %1 = tensor.expand_shape %0 [[0, 1], [2]] output_shape [%arg1, %arg2, %arg3]
+      : tensor<?x?xf32> into tensor<?x?x?xf32>
+  return %1 : tensor<?x?x?xf32>
+}
+// CHECK-LABEL: @no_fold_expand_of_collapse_dynamic
+//       CHECK:   tensor.collapse_shape
+//       CHECK:   %[[EXPAND:.+]] = tensor.expand_shape
+//       CHECK:   return %[[EXPAND]]
```
Comment (qedawkins) [inline mlir/test/Dialect/Tensor/canonicalize.mlir:1208]: Please add a test for the cases with different reassociation indices (reflecting the above comment)

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h [hunk 2]
```diff
@@ -85,21 +85,55 @@ bool isReassociationValid(ArrayRef<AffineMap> reassociation,
...
+  //      shapes are equal for each reassociation.
+  auto reassociations = reshapeOp.getReassociationIndices();
+  auto inverseReassociations = reshapeSrcOp.getReassociationIndices();
+  if (reassociations != inverseReassociations)
+    return nullptr;
+  ArrayRef<int64_t> expandedSrcShape = srcType.getShape();
+  ArrayRef<int64_t> expandedResultShape = resultType.getShape();
+  if (llvm::none_of(reassociations, [&](auto reInd) {
+        auto srcSlice = expandedSrcShape.slice(reInd.front(), reInd.size());
+        auto resSlice = expandedResultShape.slice(reInd.front(), reInd.size());
```
Comment (qedawkins) [inline mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:131]: nit: Spell out the result types.

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h [hunk 3]
```diff
@@ -85,21 +85,55 @@ bool isReassociationValid(ArrayRef<AffineMap> reassociation,
...
+        return srcSlice == resSlice &&
```
Comment (qedawkins) [inline mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:132]: This is saying `llvm::none_of` the sub-slices are equivalent, then do the folding? That seems reversed. Also you already have a check that `srcType == resultType` and `reassociations == inverseReassociations` above so this comparison should always resolve to true.

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h [hunk 4]
```diff
@@ -85,21 +85,55 @@ bool isReassociationValid(ArrayRef<AffineMap> reassociation,
...
+
+  // Fold producer-consumer reshape ops when they are perfect inverses of each
+  // other:
+  //   1) Reassociation indices are equivalent.
+  //   2) Boundary types are equivalent.
+  //   3) No reassociations have more than 1 dynamic dimension, and reassociated
+  //      shapes are equal for each reassociation.
+  auto reassociations = reshapeOp.getReassociationIndices();
+  auto inverseReassociations = reshapeSrcOp.getReassociationIndices();
+  if (reassociations != inverseReassociations)
```
Comment (qedawkins) [inline mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:125]: nit: Fold `inverseReassociations` into this check.

**Review by qedawkins — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h [hunk 5]
```diff
@@ -85,21 +85,55 @@ bool isReassociationValid(ArrayRef<AffineMap> reassociation,
...
+    return nullptr;
+  auto srcType = reshapeSrcOp.getSrcType();
+  auto resultType = reshapeOp.getResultType();
+  if (srcType != resultType)
+    return nullptr;
+
+  // If the reshapes are expanding and then collapsing, the ops can be folded
+  // despite multiple dynamic dimensions.
+  if (srcType.getRank() < reshapeSrcOp.getResultType().getRank())
+    return reshapeSrcOp.getSrc();
```
[the remark this answers is not in the document]
Comment (Max191) [reply mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h]: yeah I messed up my ordering a bit, fixed now

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h [hunk 6]
```diff
@@ -85,21 +85,51 @@ bool isReassociationValid(ArrayRef<AffineMap> reassociation,
...
+  //      shapes are equal for each reassociation.
+  auto reassociations = reshapeOp.getReassociationIndices();
+  if (reassociations != reshapeSrcOp.getReassociationIndices())
+    return nullptr;
+  // If the reshapes are expanding and then collapsing, the ops can be folded
+  // despite multiple dynamic dimensions.
+  if (srcType.getRank() < reshapeSrcOp.getResultType().getRank())
+    return reshapeSrcOp.getSrc();
+  ArrayRef<int64_t> expandedSrcShape = srcType.getShape();
+  ArrayRef<int64_t> expandedResultShape = resultType.getShape();
```
Comment (qedawkins) [inline mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:125]: nit: unused variable

**Review by qedawkins — APPROVED** [review]

**Pull request merged**